### ISSUE TITLE
Create and delete relationsship in data object

### ISSUE THREAD

**Issue by noheton** [noheton/shepard#384]
***In GitLab by @wiglef on Dec 16, 2024, 14:43 GMT+1:***

# Feature Request

## :goal_net: Goal

As a `shepard user` I need `dialogues` so that `i can create, delete and annotate relationsships in my data objects`.

## :sports_medal: Acceptance Criteria

1. there is a plus icon at the relationsship content box that allows to add relationsships :white_check_mark: 
   1. after click on add button the dialogue opens and displays a selection between "predecessor", "sucessor" and "custom" :white_check_mark:
   1. Predecessor / Successor show an input field that enables to type name or ID of the data object with autocomplete enabled :white_check_mark:
      1. save button only available as soon as data object selected. :white_check_mark:
   1. For "Custom" relationsship there is a optional input field that allows to enter any string as "Custom Relationsship type" :white_check_mark:
      1. there is a dropdown that requires the user select if the relationsship links an URL, Collection or data object :white_check_mark:
      1. for collections and data objects there is an auto-fill input field that allows to select a collection / data object to be linked :white_check_mark:
      1. Custom relationsships have a mandatory filed called "name". it shall be prefilled automatically with the name of the collection / data object selected / page title for URL. the field shall be updated when the previous field is filled and can be edited manually by user :white_check_mark:
      1. The field "relationsship" that is already modeled for Data Object and Collection relationsships shall be added to the URL relationsships as well, so that the user can name the meaning of the URL. :white_check_mark:
  
2. existing relationsships show a trash can icon in the line to delete the relationsship. Clicking so deletes the relationsship after confirmation. :white_check_mark:

## :star: Visual Design

Delete: https://www.figma.com/design/ivBPq71AuUl2ohFnrzHIKF/Visual-Design?node-id=1575-19336&m=dev 

Create: https://www.figma.com/design/ivBPq71AuUl2ohFnrzHIKF/Visual-Design?node-id=1551-18795&m=dev

## :bomb: Out of Scope
3. Data object reference, collection reference and URL reference can be annotaten and thus show a semantic annotation icon in line that leads to #420  ==> needs to be done with story #420

## :bulb: Hints

Backend needs to be extended to allow an operation to add successor directly - 

## :thinking_face: Risks & Open Questions

**Assignees:** xit-max-hey

*Migrated from GitLab: https://gitlab.com/dlr-shepard/shepard/-/work_items/384*

**Comment by noheton** [comment]
***In GitLab by @wiglef on Dec 16, 2024, 14:43 GMT+1:***

added dlr-shepard#24 as parent epic

**Comment by noheton** [comment]
***In GitLab by @wiglef on Dec 16, 2024, 14:44 GMT+1:***

marked this issue as blocked by #300

**Comment by noheton** [comment]
***In GitLab by @wiglef on Jan 22, 2025, 16:53 GMT+1:***

removed from epic &24

**Comment by noheton** [comment]
***In GitLab by @wiglef on Jan 22, 2025, 16:53 GMT+1:***

added dlr-shepard#8 as parent epic

**Comment by noheton** [comment]
***In GitLab by @wiglef on Feb 6, 2025, 16:47 GMT+1:***

Visual Design and ACs have been adapted after BLR, so that the dialogue is now simplified.

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Apr 9, 2025, 16:25 GMT+2:***

mentioned in commit 65f0a3b7195edce192c9dca283dd47ea2a97348d

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Apr 10, 2025, 13:50 GMT+2:***

1. there is a plus icon at the relationsship content box that allows to add relationsships :white_check_mark:
   1. after click on add button the dialogue opens and displays a selection between "predecessor", "sucessor" and "custom" :white_check_mark:
   2. Predecessor / Successor show an input field that enables to type name or ID of the data object with autocomplete enabled
      1. save button only available as soon as data object selected. :white_check_mark:
   3. For "Custom" relationsship there is a optional input field that allows to enter any string as "Custom Relationsship type"
      1. there is a dropdown that requires the user select if the relationsship links an URL, Collection or data object :white_check_mark: 
      2. for collections and data objects there is an auto-fill input field that allows to select a collection / data object to be linked :white_check_mark: 
      3. Custom relationsships have a mandatory filed called "name". it shall be prefilled automatically with the name of the collection / data object selected / page title for URL. the field shall be updated when the previous field is filled and can be edited manually by user
      4. The field "relationsship" that is already modeled for Data Object and Collection relationsships shall be added to the URL relationsships as well, so that the user can name the meaning of the URL.
2. existing relationsships show a trash can icon in the line to delete the relationsship. Clicking so deletes the relationsship after confirmation.

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Apr 10, 2025, 16:10 GMT+2:***

mentioned in commit 053a1b402fe7b686400fc4ec48322257094ef11e

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 14, 2025, 14:26 GMT+2:***

mentioned in commit 8c046c4e3cb9503f0ffda2dcacd4d5b91a4372ce

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 14, 2025, 14:51 GMT+2:***

mentioned in commit 24d8251008861fb8bbe0fd90173bf0c784686e93

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 15:14 GMT+2:***

mentioned in commit 9ca3e613741c7b02473137210f2ea262941252c9

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 15:14 GMT+2:***

mentioned in commit 75586414fbf1f01f47172dde0bea770ffc0e94e0

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 15:14 GMT+2:***

mentioned in commit 70a862acc4671f3d05a92079a0d9b7e10f0915a4

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 15:14 GMT+2:***

mentioned in commit a05748d055fa1bac738fe271c7b908444639aa67

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 16:05 GMT+2:***

mentioned in commit 588da4d002dc60208f64a194cce50b2b360c3eeb

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 16:43 GMT+2:***

mentioned in commit 909aa7f00131acb3c0d4fa08b838010f55a9beaa

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 16:43 GMT+2:***

mentioned in commit e057e21c91717edce7ea88e28210307ec47d80b9

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 17:30 GMT+2:***

mentioned in commit dc001f8c09581a32537430a53885b12b753f4a9c

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 17:30 GMT+2:***

mentioned in commit e31e2474c0df93d74170e08b6d514535a66dbfe5

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 18:18 GMT+2:***

mentioned in commit 65e1fad0954a5a6adc3874985f9250d4e71ba974

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 15, 2025, 18:23 GMT+2:***

mentioned in commit a2d6c9bbe444b3d0bfab9f3a31d8605bc449f7df

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 11:59 GMT+2:***

mentioned in commit 13144d41a6c06dfe796d1ba828dcb411fd47d0b1

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 11:59 GMT+2:***

mentioned in commit 4251463a94abcae12687ae95b19d356c133aaca7

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 11:59 GMT+2:***

mentioned in commit 7947d3f45a8a5462aae7f3d20d3c5f4773b7eb87

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 11:59 GMT+2:***

mentioned in commit 6a7b08c1b46ec4f911a3a223720b7025ad17eb38

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 16:46 GMT+2:***

mentioned in commit 6f86331d5003107ffa5ea7af2554bbeb719a4847

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 16:56 GMT+2:***

mentioned in commit e46a1d81d3226d0a9080249ec453cad3fa7cf4a5

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 17:17 GMT+2:***

mentioned in commit ef2a4c7df7f16b1a1f80cefeeeef0b58b14292f8

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 17:17 GMT+2:***

mentioned in commit 4c9795dd512a1f5f862ddc70ced1d184f914bef1

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 16, 2025, 17:24 GMT+2:***

:test_tube: **Testing Notes** :test_tube:

- Expand the `Relationship` expandable in the DataObject detail view, a list of relationships and a button should appear in the top right corner of the table to add new relationships
- Create a new predecessor relationship
- Create a new successor relationship
- Select 'Custom' to create custom relationships/ references
- Create a new Collection reference
- Create a new Data Object reference to a data object in the same collection
- Create a new Data Object reference to a data object in a different collection
- Create a new URI reference
  - change the default 'URI' value to something else in the 'Custom Relationship Type' field
  - this value appears in the relationship table in the most left column under 'Relationship'
- Try to click the 'Save Changes' button without having valid values in each of the required fields
- After each new addition the table should have directly been updated with the new relationship
- A green success message appears after creating a new relationship
- Delete the predecessor relationship
- Delete the successor relationship
- Delete the Collection reference
- Delete both Data Object references
- The table should now be empty (without reloading the page)

**Comment by noheton** [comment]
***In GitLab by @xit-luk-wer on Apr 16, 2025, 17:49 GMT+2:***

mentioned in commit e7e16829e97f45b3aa00e51e28f97314299622f3

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 17, 2025, 10:52 GMT+2:***

mentioned in commit 29e2474f75e821b0ea588596a4462df4d198e19b

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 17, 2025, 11:01 GMT+2:***

mentioned in commit 0f438f99658bf2f5a7d7c74235f15ee74ba6840c

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 22, 2025, 14:00 GMT+2:***

There currently is an issue with the URI reference not properly using the new 'relationship' field to display the relationship type in the dataobject's relationship table.

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 22, 2025, 14:14 GMT+2:***

mentioned in commit c6e993e0214c4669634bdf61b0c5f92947407ff0

**Comment by noheton** [comment]
***In GitLab by @xit-max-hey on Apr 22, 2025, 14:14 GMT+2:***

This has been fixed here:
c6e993e0214c4669634bdf61b0c5f92947407ff0

**Comment by noheton** [comment]
***In GitLab by @xit-mohamed on Apr 23, 2025, 16:12 GMT+2:***

Testing:

1. there is a plus icon at the relationsship content box that allows to add relationsships :white_check_mark:
   1. after click on add button the dialogue opens and displays a selection between "predecessor", "sucessor" and "custom" :white_check_mark:
   2. Predecessor / Successor show an input field that enables to type name or ID of the data object with autocomplete enabled :white_check_mark:
      1. save button only available as soon as data object selected. :white_check_mark:
   3. For "Custom" relationsship there is a optional input field that allows to enter any string as "Custom Relationsship type" :white_check_mark:
      1. there is a dropdown that requires the user select if the relationsship links an URL, Collection or data object :white_check_mark:
      2. for collections and data objects there is an auto-fill input field that allows to select a collection / data object to be linked :white_check_mark:
      3. Custom relationsships have a mandatory filed called "name". it shall be prefilled automatically with the name of the collection / data object selected / page title for URL. the field shall be updated when the previous field is filled and can be edited manually by user :white_check_mark:
      4. The field "relationsship" that is already modeled for Data Object and Collection relationsships shall be added to the URL relationsships as well, so that the user can name the meaning of the URL. :white_check_mark:
2. existing relationsships show a trash can icon in the line to delete the relationsship. Clicking so deletes the relationsship after confirmation. :white_check_mark:

Everything is functional :white_check_mark:
